File: apps/web/src/entities/task/api/tasks.api.ts
import { api } from '$shared/api/client'
import type { TaskPeriod, TaskWithPeriod, TaskLevel, TaskStatus, RecurringConfig } from '../model/task.types'

type CreateTaskDto = {
  title: string
  description?: string
  level: TaskLevel
  periodStart: string
  deadlineMonth?: number
  targetDate?: string
  scheduledTime?: string
  parentTaskId?: string
  recurringConfig?: Omit<RecurringConfig, 'id' | 'taskId' | 'createdAt'>
}

type UpdateTaskDto = Partial<{
  title: string
  description: string | null
}>

type UpdatePeriodDto = Partial<{
  status: TaskStatus
}>

export const tasksApi = {
  /** Get tasks with their current period for a given week/month/year */
  getByPeriod: (params: { periodType: TaskLevel; periodStart: string }) =>
    api.get<TaskWithPeriod[]>(`/tasks?${new URLSearchParams(params).toString()}`),

  /** Get all backlog periods */
  getBacklog: () => api.get<TaskWithPeriod[]>('/tasks?status=backlog'),

  /** Get archived */
  getArchive: () => api.get<TaskWithPeriod[]>('/tasks?status=archived'),

  create: (dto: CreateTaskDto) => api.post<TaskWithPeriod>('/tasks', dto),

  /** Update task metadata (title, description) */
  update: (taskId: string, dto: UpdateTaskDto) =>
    api.patch<TaskWithPeriod>(`/tasks/${taskId}`, dto),

  /** Toggle period status (done ↔ todo) */
  updatePeriod: (periodId: string, dto: UpdatePeriodDto) =>
    api.patch<TaskPeriod>(`/task-periods/${periodId}`, dto),

  /** Replan: create a new period for the task in the next period */
  replan: (taskId: string, dto: { periodStart: string }) =>
    api.post<TaskWithPeriod>(`/tasks/${taskId}/replan`, dto),

  remove: (taskId: string) => api.delete<void>(`/tasks/${taskId}`),

  /** Get subtasks for a task */
  getSubtasks: (taskId: string) => api.get<TaskWithPeriod[]>(`/tasks/${taskId}/subtasks`),
}

File: apps/mobile/lib/shared/api/tasks_api.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../../entities/task/task.dart';
import 'api_client.dart';

class TasksApi {
  final ApiClient _client;

  TasksApi(this._client);

  Future<List<TaskWithPeriod>> fetchTasks({String? view}) async {
    final res = await _client.get(
      '/tasks',
      queryParameters: view != null ? {'view': view} : null,
    );
    return (res.data as List).map((j) => TaskWithPeriod.fromJson(j)).toList();
  }

  Future<void> toggleTask(String periodId, TaskStatus status) async {
    await _client.patch('/task-periods/$periodId', data: {'status': status.name});
  }

  Future<TaskWithPeriod> createTask(Map<String, dynamic> data) async {
    final res = await _client.post('/tasks', data: data);
    return TaskWithPeriod.fromJson(res.data);
  }

  Future<List<TaskWithPeriod>> getSubtasks(String taskId) async {
    final res = await _client.get('/tasks/$taskId/subtasks');
    return (res.data as List).map((j) => TaskWithPeriod.fromJson(j)).toList();
  }

  Future<void> updateTask(String id, Map<String, dynamic> data) async {
    await _client.patch('/tasks/$id', data: data);
  }

  Future<void> deleteTask(String id) async {
    await _client.delete('/tasks/$id');
  }

  Future<TaskWithPeriod> replanTask(String taskId, String periodStart) async {
    final res = await _client.post('/tasks/$taskId/replan', data: {'periodStart': periodStart});
    return TaskWithPeriod.fromJson(res.data);
  }
}

final tasksApiProvider =
    Provider((ref) => TasksApi(ref.watch(apiClientProvider)));
